cpp/include/cudf/join.hpp

```diff
@@ -438,6 +446,101 @@ class hash_join {
+   * @param probe The probe table, from which the keys are probed
+   * @param compare_nulls Controls whether null join-key values should match or not
+   * @param stream CUDA stream used for device memory operations and kernel launches
+   */
+  unique_hash_join(cudf::table_view const& build,
+                   cudf::table_view const& probe,
+                   null_equality compare_nulls,
+                   rmm::cuda_stream_view stream = cudf::get_default_stream());
+
+  /**
+   * @brief Constructs a unique hash join object for subsequent probe calls
+   *
+   * @param build The build table that doesn't contain duplicate elements
+   * @param probe The probe table, from which the keys are probed
+   * @param has_nulls Flag to indicate if there exists any nulls in the `build` table or
+   *        any `probe` table that will be used later for join
+   * @param compare_nulls Controls whether null join-key values should match or not
+   * @param stream CUDA stream used for device memory operations and kernel launches
+   */
+  unique_hash_join(cudf::table_view const& build,
+                   cudf::table_view const& probe,
+                   nullable_join has_nulls,
+                   null_equality compare_nulls,
+                   rmm::cuda_stream_view stream = cudf::get_default_stream());
+
+  /**
+   * Returns the row indices that can be used to construct the result of performing
+   * an inner join between two tables. @see cudf::inner_join(). Behavior is undefined if the
+   * provided `output_size` is smaller than the actual output size.
+   *
+   * @param output_size Optional value which allows users to specify the exact output size
+   * @param stream CUDA stream used for device memory operations and kernel launches
+   * @param mr Device memory resource used to allocate the returned table and columns' device
+   * memory.
+   *
+   * @return A pair of columns [`left_indices`, `right_indices`] that can be used to construct
+   * the result of performing an inner join between two tables with `build` and `probe`
+   * as the join keys .
+   */
+  std::pair<std::unique_ptr<rmm::device_uvector<size_type>>,
+            std::unique_ptr<rmm::device_uvector<size_type>>>
+  inner_join(std::optional<std::size_t> output_size = {},
+             rmm::cuda_stream_view stream           = cudf::get_default_stream(),
+             rmm::mr::device_memory_resource* mr    = rmm::mr::get_current_device_resource()) const;
+
+  /**
+   * Returns the exact number of matches (rows) when performing an inner join
+   *
+   * @param stream CUDA stream used for device memory operations and kernel launches
+   *
+   * @return The exact number of output when performing an inner join between two tables with
+   * `build` and `probe` as the join keys .
+   */
+  // TODO: still desired?
+  [[nodiscard]] std::size_t inner_join_size(
+    rmm::cuda_stream_view stream = cudf::get_default_stream()) const;
```

Comment:
At least from the Spark perspective, we only desire the join size when trying to avoid explosive joins.  Unique joins are never explosive, so IMO the only reason to keep this is if someone wants to guarantee the gather map buffers from the unique join are exactly sized by passing the expected size computed via this API.

On a related note, I'm assuming the gather map vectors returned from the unique_hash_join join methods are not guaranteed to be "rightsized" and can have a larger underlying allocation than necessary to store the gather map.

Reply:
If this is removed, we may want to consider removing the optional output size parameter to the join method(s) since there would be no method in libcudf to help calculate that value.

Reply:
I'm mostly curious if the vectors have the possibility of being over-allocated for the reported size. I'm assuming it is only performing the join once, which means it needs to allocate the worst-case size of the gather maps before running the join code to fill the maps.  Using `rmm::device_uvector::resize` to shrink the vectors to the correct size afterwards is very fast but will leave unused memory in the result vectors, whereas `rmm::device_uvector::shrink_to_fit` will actually free the unused memory (at the cost of a copy and temporarily higher memory pressure during that copy).

There could be up to 16 gigabytes of unused memory in the extreme case of both tables being maximum size and the gather maps are empty (2\*\*31 entries of 4 bytes each per map and there's two maps).  I'm totally cool if there's the potential for over-allocated gather maps for performance.  If that's indeed the case, on the Spark side we might want to  pay for the `shrink_to_fit` afterwards to free up memory in some cases.

Reply:
I see. In that sense, we definitely over-allocate when creating the gather map vectors since they reserve the probe table size as the vector capacity.

Once the prototype works, we can benchmark the difference between `resize` and `shrink_to_fit` for performance tuning.

Reply:
I don't think it will make sense to blindly call `shrink_to_fit`, as we could copy an almost 2GB gather map to save a few bytes.  I was thinking on the Spark side we would use some kind of heuristic (e.g.: some combination of absolute threshold and percentage threshold of wasted memory) to decide whether it's worth taking the hit.  Seems like an application-side decision to make, with the downside that the API doesn't make it clear the result may be significantly over-allocated. Apps may be surprised if they don't read the docs.

Reply:
Great conversation above. I agree this feels like the caller’s responsibility and not libcudf, especially with the heuristics on memory. For most use cases I would guess the gather map is only held temporarily.